src/mame/sinclair/atm.cpp

```diff
@@ -13,118 +23,379 @@ Not working because of banking issues.
+#define LOGWARN(...)  LOGMASKED(LOG_WARN,  __VA_ARGS__)
+
+#define RAM_MASK 0x40
+#define DOS7FFD_MASK 0x80
+#define SHADOW m_beta->is_active()
+#define PEN_PAGE(bank) m_pages_map[BIT(m_port_7ffd_data, 4)][bank]
+
 class atm_state : public spectrum_128_state
 {
 public:
 	atm_state(const machine_config &mconfig, device_type type, const char *tag)
 		: spectrum_128_state(mconfig, type, tag)
-		, m_bank1(*this, "bank1")
-		, m_bank2(*this, "bank2")
-		, m_bank3(*this, "bank3")
-		, m_bank4(*this, "bank4")
+		, m_bank_view0(*this, "bank_view0")
+		, m_bank_view1(*this, "bank_view1")
+		, m_bank_view2(*this, "bank_view2")
+		, m_bank_view3(*this, "bank_view3")
+		, m_bank_rom(*this, "bank_rom%u", 0U)
 		, m_beta(*this, BETA_DISK_TAG)
+		, m_centronics(*this, "centronics")
+		, m_palette(*this, "palette")
 	{ }
 
 	void atm(machine_config &config);
 	void atmtb2(machine_config &config);
 
 protected:
-	virtual void machine_start() override;
-	virtual void machine_reset() override;
+	void machine_start() override;
+	void machine_reset() override;
+	void video_start() override;
+
+	rectangle get_screen_area() override;
+	u8 get_border_color(u16 hpos, u16 vpos) override;
+	void spectrum_update_screen(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect) override;
 
 private:
-	void atm_port_7ffd_w(uint8_t data);
-	uint8_t beta_neutral_r(offs_t offset);
-	uint8_t beta_enable_r(offs_t offset);
-	uint8_t beta_disable_r(offs_t offset);
+	u8 beta_neutral_r(offs_t offset);
+	u8 beta_enable_r(offs_t offset);
+	u8 beta_disable_r(offs_t offset);
+
+	void atm_ula_w(offs_t offset, u8 data);
+	void atm_port_ffff_w(offs_t offset, u8 data);
+	void atm_port_ff77_w(offs_t offset, u8 data);
+	void atm_port_fff7_w(offs_t offset, u8 data);
+	void atm_port_7ffd_w(offs_t offset, u8 data);
 
 	void atm_io(address_map &map);
 	void atm_mem(address_map &map);
 	void atm_switch(address_map &map);
 
+	void atm_update_video_mode();
+	void atm_update_screen_lo(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+	void atm_update_screen_hi(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+	void atm_update_screen_tx(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
 	void atm_update_memory();
 
-	required_memory_bank m_bank1;
-	required_memory_bank m_bank2;
-	required_memory_bank m_bank3;
-	required_memory_bank m_bank4;
+	memory_view m_bank_view0;
+	memory_view m_bank_view1;
+	memory_view m_bank_view2;
+	memory_view m_bank_view3;
+	required_memory_bank_array<4> m_bank_rom;
 	required_device<beta_disk_device> m_beta;
+	required_device<centronics_device> m_centronics;
+	required_device<device_palette_interface> m_palette;
+
+	int m_pen;            // PEN - extended memory manager
+	int m_cpm;
+	u8 m_pages_map[2][4]; // map: 0,1
+
+	int m_pen2;           // palette selector
+	int m_rg = 0b011;     // 0:320x200lo, 2:640:200hi, 3:256x192zx, 6:80x25txt
+	int m_br3;
 
 	address_space *m_program;
```

Comment:
You’d be better off using a `memory_access_specific` for this – it avoids some levels of indirection.

Reply:
Can you suggest an alternative? This used in many spectrum clones and I expect has extremely performance hit.

Reply:
At this point I can get rid of m_program and directly call ```m_maincpu->space(AS_PROGRAM).read_byte(offset)``` in memory switcher methods.
Would it be acceptable?

Reply:
For an example, see the Osborne 1 driver (src/mame/osborne/osborne1*.*) – it has one for the program space of the main CPU.  You can also look in a lot of CPU cores and see examples.
* Declare class member like `memory_access<16, 0, 0, ENDIANNESS_LITTLE>::specific m_program;` (@galibert can explain the numbers – they depend on the CPU’s address space configuration).
* In `machine_start`, so something like `m_maincpu->space(AS_PROGRAM).specific(m_program);` to set it up.
* Then you can do like `m_program.read_byte(...)` etc. and it will bypass some layers of indirection required when using an `address_space` directly.

Reply:
I don't know if you treat this as critical. To my excuse:
1. That's inherited from old code and didn't touched in this PR
2. I'm planning entirely revisit how shadow ROMs are activated in all Spectrums later. Now I'm thinking to use TAPs for the switch but need perform some experiments with it.

Reply:
OK. I have with specific now. After investigating TAP have so many ideas - will do it later.